EngineLayer/FdrAnalysis/PEPValueAnalysisGeneric.cs

```diff
@@ -57,84 +57,150 @@ public static string ComputePEPValuesForAllPSMsGeneric(List<PeptideSpectralMatch
-            //     (train or test). This can be used to ensure no label leakage from the train to
-            //     the test set. If null no row grouping will be performed.
-            //
-            //   seed:
-            //     Seed for the random number generator used to select rows for the train-test split.
-            //     The seed, '42', is not random but fixed for consistancy. According to the supercomputer Deep Thought the answer to the question of life, the universe and everything was 42 (in Douglas Adam’s Hitchhikers Guide to the Galaxy).
-
-            int randomSeed = 42;
-
-            double fraction = Math.Min(0.25, 1000000.0 / (double)psms.Count);//make training set 25% of the data up to a training set of one million psms
-            TrainTestData trainTestSplit = mlContext.Data.TrainTestSplit(dataView, testFraction: fraction, null, randomSeed);
-            IDataView trainingData = trainTestSplit.TrainSet;
-            IDataView testData = trainTestSplit.TestSet;
+            //the number of groups used for cross-validation is hard-coded at four. Do not change this number without changes other areas of effected code. 
+            const int numGroups = 4;
+
+            List<int>[] psmGroupIndices = Get_PSM_Group_Indices(psms.Count, numGroups);
+            IEnumerable<PsmData>[] PSMDataGroups = new IEnumerable<PsmData>[numGroups];
+            for (int i = 0; i < numGroups; i++)
+            {
+                PSMDataGroups[i] = CreatePsmData(searchType, fileSpecificParameters, psms, psmGroupIndices[i], sequenceToPsmCount, fileSpecificTimeDependantHydrophobicityAverageAndDeviation_unmodified, fileSpecificTimeDependantHydrophobicityAverageAndDeviation_modified, fileSpecificMedianFragmentMassErrors, chargeStateMode);
+            }
+
+            TransformerChain<BinaryPredictionTransformer<Microsoft.ML.Calibrators.CalibratedModelParametersBase<Microsoft.ML.Trainers.FastTree.FastTreeBinaryModelParameters, Microsoft.ML.Calibrators.PlattCalibrator>>>[] trainedModels = new TransformerChain<BinaryPredictionTransformer<Microsoft.ML.Calibrators.CalibratedModelParametersBase<Microsoft.ML.Trainers.FastTree.FastTreeBinaryModelParameters, Microsoft.ML.Calibrators.PlattCalibrator>>>[numGroups];
+
             var trainer = mlContext.BinaryClassification.Trainers.FastTree(labelColumnName: "Label", featureColumnName: "Features", numberOfTrees: 400);
             var pipeline = mlContext.Transforms.Concatenate("Features", trainingVariables)
                 .Append(trainer);
-            var trainedModel = pipeline.Fit(trainingData); //this is NOT thread safe
 
-            int maxThreads = fileSpecificParameters.FirstOrDefault().fileSpecificParameters.MaxThreadsToUsePerFile;
-            object lockObject = new object();
+            List<CalibratedBinaryClassificationMetrics> allMetrics = new List<CalibratedBinaryClassificationMetrics>();
+            int sumOfAllAmbiguousPeptidesResolved = 0;
+            try
```

Comment:
This is just a pet peeve, but I hate try-catch statements. They dramatically slow down the code.
If you can think of an if-else to handle it instead, that'd be preferred. No worries if not.